fuse_core/include/fuse_core/graph_deserializer.h

```diff
@@ -0,0 +1,102 @@
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_CORE_GRAPH_DESERIALIZER_H
+#define FUSE_CORE_GRAPH_DESERIALIZER_H
+
+#include <fuse_msgs/SerializedGraph.h>
+#include <fuse_core/constraint.h>
+#include <fuse_core/graph.h>
+#include <fuse_core/variable.h>
+#include <pluginlib/class_loader.h>
+
+
+namespace fuse_core
+{
+
+/**
+ * @brief Serialize a graph into a message
+ */
+void serializeGraph(const fuse_core::Graph& graph, fuse_msgs::SerializedGraph& msg);
+
+/**
+ * @brief Deserialize a graph
+ *
+ * The deserializer object loads all of the known Variable and Constraint libraries, allowing derived types contained
+ * within the graph to be properly deserialized. The libraries will be unloaded on destruction. As a consequence, the
+ * deserializer object must outlive any created graph instances.
+ */
+class GraphDeserializer
+{
+public:
+  /**
+   * @brief Constructor
+   */
+  GraphDeserializer();
+
+  /**
+   * @brief Destructor
+   */
+  ~GraphDeserializer();
+
+  /**
+   * @brief Deserialize a SerializedGraph message into a fuse Graph object.
+   *
+   * If no plugin is available for a contained Variable or Constraint, or an error occurs during deserialization,
+   * an exception is thrown.
+   *
+   * @param[IN]  msg  The SerializedGraph message to be deserialized
+   * @return          A unique_ptr to a derived Graph object
+   */
+  fuse_core::Graph::UniquePtr deserialize(const fuse_msgs::SerializedGraph::ConstPtr& msg);
```

Comment:
I need to test, but I don't _think_ that would work. A static method would allow me to call `deserialize()` without first creating an object. I believe I need to load the shared libraries that contain the variables and constraints before deserializing the graph. At least, that how it works in my head.

Reply:
Yes, that's true, but there's an issue right now. You should have a singleton to load the shared libraries only once, and more importantly, unload them only once.

Currently, `fuse_echo` crashes on shutdown because of that, with this message:
``` bash
terminate called after throwing an instance of 'class_loader::LibraryUnloadException'
  what():  Attempt to unload library that class_loader is unaware of.
```

I commented out the transaction deserializer code in `fuse_echo` and it didn't crash, so I'm quite sure the issue is that when the graph deserializer (call after the transaction deserializer because it's before in the attributes list) is destroyed it unloads the shared libraries that has already been unloaded by the transaction desrializer. Indeed, the core dump I get shows the crash coming from the graph deserializer destructor, on `pluginlib` unloading methods.

Reply:
@ayrton04 I was proposing a solution for this, but from the last comments from @svwilliams it seems that the issue was bigger than I initially thought.

Reply:
Gotcha. My first comment to you was about this comment you made originally:

> @ayrton04 I wonder if the SerializedPublisher should also have a device_name (and device_id), as the Odometry2DPublisher 🤔

What would be the advantage of that?